src/micro.c

```diff
@@ -0,0 +1,680 @@
+        }
+    }
+
+    micro_free_cloned_service_config(m->cfg);
+    natsConn_release(m->nc);
+    natsMutex_Destroy(m->service_mu);
+    NATS_FREE(m);
+}
+
+static inline microError *new_service_config(microServiceConfig **ptr)
+{
+    *ptr = NATS_CALLOC(1, sizeof(microServiceConfig));
+    return (*ptr == NULL) ? micro_ErrorOutOfMemory : NULL;
+}
+
+microError *
+micro_clone_service_config(microServiceConfig **out, microServiceConfig *cfg)
+{
+    microError *err = NULL;
+    microServiceConfig *new_cfg = NULL;
+
+    if (out == NULL || cfg == NULL)
+        return micro_ErrorInvalidArg;
+
+    err = new_service_config(&new_cfg);
+    if (err == NULL)
+    {
+        memcpy(new_cfg, cfg, sizeof(microServiceConfig));
+    }
+    // the strings are declared const for the public, but in a clone these need
+    // to be duplicated.
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Name, cfg->Name));
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Version, cfg->Version));
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Description, cfg->Description));
+    MICRO_CALL(err, micro_clone_endpoint_config(&new_cfg->Endpoint, cfg->Endpoint));
+    if (err != NULL)
+    {
+        micro_free_cloned_service_config(new_cfg);
+        return err;
+    }
+
+    *out = new_cfg;
+    return NULL;
+}
+
+void micro_free_cloned_service_config(microServiceConfig *cfg)
+{
+    if (cfg == NULL)
+        return;
+
+    // the strings are declared const for the public, but in a clone these need
+    // to be freed.
+    NATS_FREE((char *)cfg->Name);
+    NATS_FREE((char *)cfg->Version);
+    NATS_FREE((char *)cfg->Description);
+    micro_free_cloned_endpoint_config(cfg->Endpoint);
+    NATS_FREE(cfg);
+}
+
+static void
+on_connection_closed(natsConnection *nc, void *closure)
+{
+    microService *m = (microService *)closure;
+    if (m == NULL)
+        return;
+
+    microService_Stop(m);
+
+    if (m->prev_on_connection_closed != NULL)
+    {
+        (*m->prev_on_connection_closed)(nc, m->prev_on_connection_closed_closure);
+    }
+}
+
+static void
+on_connection_disconnected(natsConnection *nc, void *closure)
+{
+    microService *m = (microService *)closure;
+
+    if (m == NULL)
+        return;
+
+    // <>/<> TODO: Should we stop the service? Not 100% how the Go client does
+    // it.
+    microService_Stop(m);
+
+    if (m->prev_on_connection_disconnected != NULL)
+    {
+        (*m->prev_on_connection_closed)(nc, m->prev_on_connection_disconnected_closure);
+    }
+}
+
+static void
+on_error(natsConnection *nc, natsSubscription *sub, natsStatus s, void *closure)
+{
+    microService *m = (microService *)closure;
+    microEndpoint *ep = NULL;
+    microError *err = NULL;
+    bool our_subject = false;
+
+    if ((m == NULL) || (sub == NULL))
+        return;
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; (!our_subject) && (ep != NULL); ep = ep->next)
+    {
+        if (micro_match_endpoint_subject(ep->subject, sub->subject))
```

Comment:
Maybe would declare a `const char*` at the top and call `natsSubscription_GetSubject()` instead of directly referencing `sub->subject` here without sub's lock.

Reply:
done